### ISSUE TITLE
REPL warnings reporting inconsistent or missing

### ISSUE THREAD

**Issue by som-snytt** [scala/bug#11957]
## reproduction steps

```scala
scala 2.11.12> @deprecated def f = 42
warning: there was one deprecation warning; re-run with -deprecation for details
f: Int

scala 2.11.12> :warnings
<console>:5: warning: method f is deprecated
  lazy val $result = f
                                          ^

scala 2.11.12> :replay -deprecation
Replaying: @deprecated def f = 42
<console>:11: warning: @deprecated now takes two arguments; see the scaladoc.
               @deprecated def f = 42
                ^
f: Int


scala 2.11.12>
```

```scala
scala 2.12.10> @deprecated def f = 42
<console>:11: warning: @deprecated now takes two arguments; see the scaladoc.
               @deprecated def f = 42
                ^
f: Int

scala 2.12.10> :warnings
Can't find any cached warnings.

scala 2.12.10> :replay -deprecation
Replaying: @deprecated def f = 42
<console>:11: warning: @deprecated now takes two arguments; see the scaladoc.
               @deprecated def f = 42
                ^
f: Int


scala 2.12.10>
```

```scala
scala 2.13.1> @deprecated def f = 42
f: Int

scala 2.13.1> :warnings
Can't find any cached warnings.

scala 2.13.1> :replay -deprecation
replay> @deprecated def f = 42
f: Int


scala 2.13.1> f
              ^
              warning: method f is deprecated
res0: Int = 42

scala 2.13.1>
```

## problem

Warnings are not always displayed.

## expectation

No idea. When `IMain#updateRecentWarnings` was introduced for 2.10, the idea was to show warnings once instead of twice, so the final compilation was suppressed, and `:warnings` is available to show anything you might have missed. REPL was refactored with more finicky compilation, so it should be recalibrated to show warnings again. Should the result line just report "there were warnings, use :warnings to display them"? assuming that command works. Or always show warnings unless `-nowarn`?

It's necessary to distinguish parsing messages, messaging on user code, and messaging on templated (wrapped) code.

**Comment by dwijnand** [comment]
This shipped in 2.13.3.

**Comment by som-snytt** [comment]
I still don't know what `:warnings` does.
